### PR TITLE
Fix --disable-strict-error-checking and -W... options handling

### PR DESCRIPTION
**Pull request by rousskov**
Commit 8b082ed (Activate extra compiler checks):

* started adding -Werror in --disable-strict-error-checking builds
* stopped adding -Wall to CXXFLAGS
* forgot that -Werror is spelled $squid_cv_cc_option_werror
* forgot that "=" is converted to "_" in $squid_cv_cc_arg... name
* forgot that "W" is present in $squid_cv_cc_arg... name
* checked squid_cv_cc_... name but set squid_cv_cc_arg... name
* used warning-generating SQUID_CC_CHECK_ARGUMENT code for testing
  compiler support for new -W options (while also using -Werror)

Correctly detecting support for a compiler warning option is difficult:

* Some compilers (e.g., clang) exit successfully when given an
  unsupported command line option unless we also give -Werror. With
  those compilers, we must use "-Werror -Wx" to test for -Wx support.
  However, that does not mean we should add -Werror to all Squid builds.

* Adding -Werror does not work for detecting supported GCC warnings: GCC
  needs -Werror=no-x to fail on unsupported -Wno-x warnings.

* We do not even know how to detect (via compiler exit code) supported
  warnings for compilers other than clang and GCC. We still add -Werror
  to all compilers other than GCC when testing -X, but, based on quick
  godbolt.org tests, -Werror does not make icc fail on unsupported -Wx.

Also delayed adding warnings until we are done adding such C++ compiler
options as -std=c++11 (and, for Irix, optimization level) because those
options may affect the set of warnings supported by the compiler. We
also (now) need SQUID_CC_GUESS_VARIANT and SQUID_CC_GUESS_OPTIONS. We
might need AC_PROG_CPP. All those things are settled after the official
code attempted to add warnings. The best code location probably needs
more work, but at least all warnings are handled in the same code now.

These changes also fix Bug 5167 (GCC -Wno-unused-private-field
miscategorized as supported).

Also fixed clang (-Wall => -Wmost) warnings exposed by the above fixes.

### CODE REVIEW

**Review by yadij — CHANGES_REQUESTED** [review]

File: src/http/RegisteredHeaders.h [hunk 1]
```diff
@@ -151,7 +151,8 @@ enum HdrKind {
 };
 
 /* POD for HeaderTable */
-struct HeaderTableRecord {
+class HeaderTableRecord {
+public:
```
Comment (kinkie) [inline src/http/RegisteredHeaders.h:155]: This is out of scope

File: src/security/Session.cc [hunk 2]
```diff
@@ -383,10 +383,10 @@ Security::SetSessionCacheCallbacks(Security::ContextPointer &ctx)
 }
 #endif /* USE_OPENSSL */
 
+#if USE_OPENSSL
```
Comment (kinkie) [inline src/security/Session.cc:386]: out of scope

**Review by kinkie — CHANGES_REQUESTED** [review]
Some out of scope changes

File: acinclude/compiler-flags.m4 [hunk 3]
```diff
@@ -113,18 +113,44 @@ AC_DEFUN([SQUID_CC_GUESS_VARIANT], [
...
 
-dnl argument is a compiler flag. It will be attempted, and if suppported
-dnl it will be added to SQUID_CXXFLAGS in the same order as calls to the macro
-AC_DEFUN([SQUID_CC_ADD_CXXFLAG_IF_SUPPORTED],[
-  SQUID_CC_ADD_CXXFLAG_IF_SUPPORTED_INTERNAL($1,m4_bpatsubst(m4_tolower([squid_cv_cc_arg$1]),[[^a-zA-Z0-9_]], [_]))
+dnl The argument is a compiler warning option (e.g. -Wall). If linking a
+dnl warning-free program while using the given warning succeeds, then the
+dnl option is added to SQUID_CXXFLAGS in the same order as calls to the macro.
+AC_DEFUN([SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED],[
+  SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED_($1,m4_bpatsubst(m4_tolower([squid_cv_cc_arg$1]),[[^a-zA-Z0-9_]], [_]))
```
[the remark this answers is not in the document]
Comment (rousskov) [reply acinclude/compiler-flags.m4]: > Whats with the rename for these macros?

These are completely different macros: Removed macros were for all compiler flags. New macros are for compiler warnings only; they are not meant for other compiler flags/options. This scope change is important enough to be reflected in the macro name IMO.

FWIW, I am also unhappy with the `_SUPPORTED` part which promises more than these macros can deliver. The macro description no longer makes that promise, but I also wanted to use something like `_SEEMS_SUPPORTED` or `_APPEARS_TO_WORK` in the macro name. I did not do it because I was worried abut the macro names hitting some (unknown to me) length limit and getting into costly debates about names. Better names are always welcome, of course.

File: acinclude/compiler-flags.m4 [hunk 4]
```diff
@@ -113,18 +113,44 @@ AC_DEFUN([SQUID_CC_GUESS_VARIANT], [
...
+dnl SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED helper
+dnl $1 is a compiler warning option (e.g., -Wall).
+dnl $2 is a "squid_cv_cc_arg<snake_case_warning_name_equivalent>" string.
+AC_DEFUN([SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED_],[
+  AC_REQUIRE([SQUID_CC_GUESS_VARIANT])
   SQUID_STATE_SAVE([CXXARGTEST])
   CXXFLAGS="$CXXFLAGS $SQUID_CXXFLAGS"
-  SQUID_CC_CHECK_ARGUMENT([$2],[$1])
+  case "$squid_cv_compiler" in
+    gcc)
```
[the remark this answers is not in the document]
Comment (rousskov) [reply acinclude/compiler-flags.m4]: Done in 0a56704.

File: acinclude/compiler-flags.m4 [hunk 5]
```diff
@@ -113,18 +113,44 @@ AC_DEFUN([SQUID_CC_GUESS_VARIANT], [
...
+      # Testing with -Werror -Wfoobar does not work well because GCC ignores
+      # unsupported _negative_ options, so we test with -Werror=foobar instead
+      # (where "foobar" is a name of a warning that may be given to us in
+      # positive -Wfoobar or negative -Wno-foobar form).
+      SQUID_CC_CHECK_ARGUMENT([$2],m4_bpatsubst([$1],[^-W], [-Werror=]))
```
[the remark this answers is not in the document]
Comment (rousskov) [reply acinclude/compiler-flags.m4]: Done in 5637af3, here and in the old line where this was copied from. That old line was already in the diff due to other necessary changes.

**Review by rousskov — COMMENTED** [review]
I believe all change requests have been addressed as of branch commit 5637af3.

**Review by yadij — APPROVED** [review]

**Comment by rousskov** [comment]
@kinkie, just a gentle reminder that this PR is waiting for your re-review. Your first review flagged required build fixes as out of scope. I assume that was due to some misunderstanding, but I cannot make further progress without your help.

**Comment by kinkie** [comment]
> @kinkie, just a gentle reminder that this PR is waiting for your re-review. Your first review flagged required build fixes as out of scope. I assume that was due to some misunderstanding, but I cannot make further progress without your help.

on it now

**Review by kinkie — APPROVED** [review]
My doubts have been answered. Good to go
